### PR TITLE
OCPBUGS-3499: UPSTREAM: <carry>: Add validation plugin for CRD-based route parity.

### PR DESCRIPTION
**Pull request by benluddy**

### CODE REVIEW

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	return routevalidation.ValidateRoute(routeObj)
+}
+
+func (routeV1) ValidateUpdate(obj runtime.Object, oldObj runtime.Object) field.ErrorList {
+	routeObj, errs := toRoute(obj)
+	if len(errs) > 0 {
+		return errs
+	}
+
+	routeOldObj, errs := toRoute(obj)
```
Comment (tkashem) [inline openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:65]: typo, should it be `oldObj`?

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	return routevalidation.ValidateRouteUpdate(routeObj, routeOldObj)
+}
+
+func (c routeV1) ValidateStatusUpdate(obj runtime.Object, oldObj runtime.Object) field.ErrorList {
+	routeObj, errs := toRoute(obj)
+	if len(errs) > 0 {
+		return errs
+	}
+
+	routeOldObj, errs := toRoute(obj)
```
Comment (tkashem) [inline openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:79]: typo, should it be `oldObj`?

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+			},
+			map[schema.GroupVersionKind]customresourcevalidation.ObjectValidator{
+				routev1.GroupVersion.WithKind("Route"): routeV1{},
+			})
+	})
+}
+
+func toRoute(uncastObj runtime.Object) (*routev1.Route, field.ErrorList) {
+	if uncastObj == nil {
+		return nil, nil
```
Comment (tkashem) [inline openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:33]: nit: i know it's never gong to be `nil` in real-life, do you think we should return an appropriate `field.ErrorList`, or is there a case for `nil`?

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go (hunk 1 above)
Comment (benluddy) [reply openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go]: Thanks!

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go (hunk 3 above)
Comment (benluddy) [reply openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:33]: I see it as harmless, and it's consistent with (i.e., copy-pasted from) comparable functions in a number of other CR validators. Do you think it's worth changing here?

Comment (tkashem) [reply openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:33]: no, let's keep it as is.

**Review by soltysh — APPROVED** [review]
/approve
/remove-label backports/unvalidated-commits
/label backports/validated-commits

**Pull request merged**